servers/rendering/renderer_viewport.cpp

```diff
@@ -526,6 +526,67 @@ void RendererViewport::_draw_viewport(Viewport *p_viewport) {
 	}
 }
 
+bool RendererViewport::_check_viewport_visible(Viewport *p_viewport, const uint64_t p_pass, Ref<XRInterface> &p_xr_interface) {
+	if (p_viewport->last_pass == p_pass && p_pass) {
+		return p_viewport->visible; // Returns the computed cached value.
+	}
+
+	p_viewport->last_pass = p_pass;
+	p_viewport->visible = false;
+
+	if (p_viewport->update_mode == RS::VIEWPORT_UPDATE_DISABLED) {
+		return false;
+	}
+
+	if (!p_viewport->render_target.is_valid()) {
+		return false;
+	}
+	//ERR_CONTINUE(!p_viewport->render_target.is_valid());
+
+	p_viewport->visible = p_viewport->viewport_to_screen_rect != Rect2();
+
+	if (p_viewport->use_xr) {
+		if (p_xr_interface.is_valid()) {
+			// Override our size, make sure it matches our required size and is created as a stereo target
+			Size2 xr_size = p_xr_interface->get_render_target_size();
+
+			// Would have been nice if we could call viewport_set_size here,
+			// but alas that takes our RID and we now have our pointer,
+			// also we only check if view_count changes in render_target_set_size so we need to call that for this to reliably change
+			p_viewport->occlusion_buffer_dirty = p_viewport->occlusion_buffer_dirty || (p_viewport->size != xr_size);
+			p_viewport->size = xr_size;
+			uint32_t view_count = p_xr_interface->get_view_count();
+			RSG::texture_storage->render_target_set_size(p_viewport->render_target, p_viewport->size.x, p_viewport->size.y, view_count);
+
+			// Inform xr interface we're about to render its viewport, if this returns false we don't render
+			p_viewport->visible = p_xr_interface->pre_draw_viewport(p_viewport->render_target);
+		} else {
```

Comment:
Hmm, while extremely unlikely, if an XR viewport is the parent viewport for a viewport set to `RS::VIEWPORT_UPDATE_WHEN_PARENT_VISIBLE`, this would cause `pre_draw_viewport` to be called prematurely and multiple times per frame. That could have unwanted consequences.

Reply:
It doesn't check multiple times per pass unless `p_pass == 0`. 
```
if (p_viewport->last_pass == p_pass && p_pass) {
	return p_viewport->visible; // Returns the computed cached value.
}
```
The above code can avoid this problem, it will directly use the cached result.